### ISSUE TITLE
Non-Bidirectional traffic on E&C bus

### ISSUE THREAD

**Issue by JoshYourITGuy** [qwertychouskie/GM-EandC-Cemu-v2#1]
[Sent packet      0x00000000 00C3108.txt](https://github.com/qwertychouskie/GM-EandC-Cemu-v2/files/15378948/Sent.packet.0x00000000.00C3108.txt)

Followed the guide provided on Reddit, using Stuart's schematic.

Able to see packets in serial monitor sent to the bus, and able to see traffic on the bus from the vehicle, but the vehicle seems to ignore traffic.

In the attached log snippet, looking at the last reset of the Pico, the Pic sends packets out, but no reply is received.
The unknown packets coming back at the end of the log are from me pressing the Source button on the head unit.
The 3-30-153 is the car asking "Is Module Present?" and not receiving anything back from the Pico.

I will be doing more testing including notating logs as to what buttons cause what messages at a later time, but this is a jumping off point for the non-bidirectional issue.

Thank you.

Side note, vehicle is a '00 Buick Park Avenue Ultra with the Concert Sound III head unit that has a CD/tape combo on board.
Linking to the E&C bus via the 12-CD disc changer plug in the trunk.

**Comment by qwertychouskie** [comment]
I'm trying to remember what the issue was when I ran into this. I'm pretty sure it's something to do with the circuit though.  It seems for you that RX is working, but TX is not.  Perhaps the transistor is the wrong type?  (NPN vs PNP)

**Comment by qwertychouskie** [comment]
Or the simple things first: are you sure you are connected to the right TX pin on the ardruino?

**Comment by JoshYourITGuy** [comment]
Using a pair of 2N3904's, rebuilt the resistor/transitor circuit, triple checked wiring, verified pinouts, TX on 22, RX on 19.

Also tried switching the voltage from the 3.3v rail up to the 5v rail, same results.

Did not add the 12v pull-up because the OEM head unit seems to have one, otherwise the bus would be dead.

I am lacking a good way to verify the Pico is transmitting.

My next thought is to duplicate the entire circuit (minus the DAC) and have them both on the bus, in theory if the Pico is transmitting on the bus, the other pico should receive that packet as well.

**Comment by JoshYourITGuy** [comment]
Duplicated the circuit and had them both on the same bus, neither one received a packet from the other, so it seems like there is an issue with TX in the circuit or the code.

Any further thoughts?

**Comment by qwertychouskie** [comment]
At this point, I'd write a simple sketch that just turns on/off the TX pin every second, and check the output with a multimeter.  It sounds like a circuit issue of some sort, and at that point you usually want to make things as easy to debug as possible.

(Is it possible you have the wrong pin configured for TX?  If I remember correctly, the pins number in software and the pin numbers on the board may or may not match up depending on what Arduino or Arduino-alike you are using)

**Comment by qwertychouskie** [comment]
Are you using a [ESP32-PICO-KIT](https://www.mouser.com/ProductDetail/356-ESP32-PICO-KIT) or something else?

**Comment by JoshYourITGuy** [comment]
ESP32-PICO-KIT-1 Development Board
says ESP32-PICO-V3 on the main chip, looks identical to your Mouser link, but the silkscreen pinout is not matching your link.
Good idea on pulsing the TX pin, we know we have the correct RX pin, let me see if I can verify the TX via a simple blink sketch to verify the numbering is correct.

**Comment by JoshYourITGuy** [comment]
Just did a simple blink sketch to pin 22, and it correctly pulsed the pin at the PICO. 0-3.3V
Checked at the E&C bus line with 12V being supplied to the circuit via bench power, and I can see it pulsing between 0 and 6mV, so potentially an issue with the circuit?

**Comment by qwertychouskie** [comment]
Yeah sounds like a circuit issue.  Double-check the values of all your resistors with a multimeter.

![circuit](https://stuartschmitt.com/e_and_c_bus/TTL_interface.png)

Note that R2 is 470 ohms, while R3 is 47 **kilo**ohms, easy to mix up.

**Comment by JoshYourITGuy** [comment]
Just rebuilt it on a new board, but used BC337's this time and metered out all resistors.
Same :(

**Comment by qwertychouskie** [comment]
> Just rebuilt it on a new board, but used BC337's this time and metered out all resistors. Same :(

Hmm, is this hooked up to the car's bus, or with a 12v line hooked up to the output of Q2 through a 22kiloohm resistor?

**Comment by JoshYourITGuy** [comment]
![2024-06-04_06h14_15](https://github.com/qwertychouskie/GM-EandC-Cemu-v2/assets/97166679/fb941d60-6195-4249-9794-51472b56f25b)
Well.... the plot thickens....
I was testing it on the car's bus, which by Stuart's account should contain the pullup in the radio (OEM GM head unit) as he indicates the bus would be dead without it.
But, when I add a pullup on one and use the bench power supply, I can see bi-directional traffic between both identical circuits when resetting them with their bus lines tied together.

Up next will be trying the one I added the pullup to in the car to see what happens.

**Comment by JoshYourITGuy** [comment]
![Screenshot 2024-06-05 100716](https://github.com/qwertychouskie/GM-EandC-Cemu-v2/assets/97166679/63fa6a3a-3f52-4343-a108-9d680972b258)
Tested with the 12v pullup resistor on one of the boards with both on the car's bus. The boards see eachother, and the boards see the car, but the car doesnt seem to respond to the boards.

Next step is to try changing the code to send other commands like CD changer or OnStar commands from the boards tp see what the car does.

**Comment by qwertychouskie** [comment]
I find it very strange that the 12v pullup doesn't seem to be present.  Perhaps the Buick radio just behaves differently.  Worst case scenario, if the physical size is the same, you could probably get a Pontiac radio cheap on eBay.

**Comment by qwertychouskie** [comment]
Looking at some side-by-sides on eBay, the mounts look at least very similar.  That said, if you can get the radio to see the Arduino as a CD changer, that would be ideal.  At least https://stuartschmitt.com/e_and_c_bus/messages.html seems to have good documentation of CD changer commands, so as long as they aren't different on your unit, it shouldn't be Too Hard™ to implement the support.
